### PR TITLE
Fix `cuda::atomic<Integral>`'s `fetch_max/fetch_min` extension

### PR DESCRIPTION
**Pull request by wmaxey**
There was previously no testing, and this has been broken since 1.3.0. I am unable to test back further.

### CODE REVIEW

File: .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+struct TestFn {
+  __host__ __device__
+  void operator()() const {
+    // Test greater
+    {
+        typedef cuda::atomic<T> A;
+        Selector<A, constructor_initializer> sel;
+        A & t = *sel.construct();
+        t = T(1);
+        assert(t.fetch_max(2) == T(1));
```
Comment (wmaxey) [inline .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp:32]: Is this expected? I believe the tests should operate similarly to `fetch_add` where the result is the current value of the atomic.

File: libcxx/include/support/atomic/atomic_cuda.h [hunk 2]
```diff
@@ -396,7 +396,7 @@ __host__ __device__
     _Tp __expected = __cxx_atomic_load(__a, memory_order_relaxed);
     _Tp __desired = __expected < __val ? __expected : __val;
 
-    while(__desired != __val &&
+    while(__desired == __val &&
```
Comment (wmaxey) [inline libcxx/include/support/atomic/atomic_cuda.h:399]: Either I've introduced a bug or there is something else wrong here. This was different from the implementation for max.

File: .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp (hunk 1 above)
Comment (griwes) [reply .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp:32]: This looks right to me.

File: libcxx/include/support/atomic/atomic_cuda.h [hunk 3]
```diff
@@ -376,6 +376,34 @@ __host__ __device__
     )
 }
 
+template <typename _Tp, typename _Delta, int _Sco, bool _Ref>
+__host__ __device__
+ _Tp __cxx_atomic_fetch_max(__cxx_atomic_base_heterogeneous_impl<_Tp, _Sco, _Ref> volatile* __a, _Delta __val, memory_order __order) {
+    _Tp __expected = __cxx_atomic_load(__a, memory_order_relaxed);
```
Comment (griwes) [inline libcxx/include/support/atomic/atomic_cuda.h:382]: This needs to be scoped, no?

File: libcxx/include/support/atomic/atomic_cuda.h [hunk 4]
```diff
@@ -376,6 +376,34 @@ __host__ __device__
...
+    _Tp __desired = __expected > __val ? __expected : __val;
+
+    while(__desired == __val &&
+            !__cxx_atomic_compare_exchange_strong(__a, &__expected, __desired, __order, __order)) {
```
Comment (griwes) [inline libcxx/include/support/atomic/atomic_cuda.h:386]: Same here (and same in fetch_min).

File: libcxx/include/support/atomic/atomic_cuda.h (hunk 3 above)
Comment (wmaxey) [reply libcxx/include/support/atomic/atomic_cuda.h:382]: `_Sco` is carried by `__a` for the `__cxx_atomic_` layer.

File: .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp (hunk 1 above)
Comment (wmaxey) [reply .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp:32]: If that's the case then the documentation and example is wrong.

https://nvidia.github.io/libcudacxx/extended_api/synchronization_primitives/atomic/fetch_max.html

File: libcxx/include/support/atomic/atomic_cuda.h (hunk 3 above)
Comment (griwes) [reply libcxx/include/support/atomic/atomic_cuda.h:382]: Ah! that makes a lot of sense.

File: .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp (hunk 1 above)
Comment (griwes) [reply .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp:32]: The docs and the example suggest that this isn't an RMW operation, but then this wouldn't be very useful. The example is also clearly not how `atom.{min,max}` work in PTX.

**Review by griwes — APPROVED** [review]
Approve, we can fix the docs independently.

File: .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp (hunk 1 above)
Comment (wmaxey) [reply .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp:32]: I hadn't dug deep enough on this. I bypassed the PTX enhancements. Will append another commit to fix.

**Review by griwes — APPROVED** [review]

**Pull request merged**